pkg/cidr/cidr_test.go

```diff
@@ -189,3 +189,112 @@ func (t *CidrTestSuite) TestEqual(c *check.C) {
 	}
 }
+
+func mustNewCIDRs(cidrs ...string) []*net.IPNet {
+	ipnets := make([]*net.IPNet, 0, len(cidrs))
+	for _, cidr := range cidrs {
+		_, ipNet, err := net.ParseCIDR(cidr)
+		if err != nil {
+			panic(err)
+		}
+		ipnets = append(ipnets, ipNet)
+	}
+	return ipnets
+}
+
+func (t *CidrTestSuite) TestRemoveAll(c *check.C) {
+	type args struct {
+		ipNets   []*net.IPNet
+		toRemove []*net.IPNet
+	}
+	tests := []struct {
+		name string
+		args args
+		want []*net.IPNet
+	}{
+		{
+			name: "remove head",
+			args: args{
+				ipNets:   mustNewCIDRs("10.10.0.0/24", "10.10.1.0/24", "10.10.2.0/24"),
+				toRemove: mustNewCIDRs("10.10.0.0/24"),
+			},
+			want: mustNewCIDRs("10.10.1.0/24", "10.10.2.0/24"),
+		},
+		{
+			name: "remove middle",
+			args: args{
+				ipNets:   mustNewCIDRs("10.10.0.0/24", "10.10.1.0/24", "10.10.2.0/24"),
+				toRemove: mustNewCIDRs("10.10.1.0/24"),
+			},
+			want: mustNewCIDRs("10.10.0.0/24", "10.10.2.0/24"),
+		},
+		{
+			name: "remove tail",
+			args: args{
+				ipNets:   mustNewCIDRs("10.10.0.0/24", "10.10.1.0/24", "10.10.2.0/24"),
+				toRemove: mustNewCIDRs("10.10.2.0/24"),
+			},
+			want: mustNewCIDRs("10.10.0.0/24", "10.10.1.0/24"),
+		},
+		{
+			name: "remove all",
+			args: args{
+				ipNets:   mustNewCIDRs("10.10.0.0/24", "10.10.1.0/24", "10.10.2.0/24"),
+				toRemove: mustNewCIDRs("10.10.0.0/24", "10.10.1.0/24", "10.10.2.0/24"),
+			},
+			want: []*net.IPNet{},
+		},
+		{
+			name: "remove none",
+			args: args{
+				ipNets:   mustNewCIDRs("10.10.0.0/24", "10.10.1.0/24", "10.10.2.0/24"),
+				toRemove: []*net.IPNet{},
+			},
+			want: mustNewCIDRs("10.10.0.0/24", "10.10.1.0/24", "10.10.2.0/24"),
+		},
+		{
+			name: "remove duplicates",
+			args: args{
+				ipNets:   mustNewCIDRs("10.10.0.0/24", "10.10.1.0/24", "10.10.2.0/24", "10.10.3.0/24", "10.10.0.0/24"),
+				toRemove: mustNewCIDRs("10.10.0.0/24", "10.10.2.0/24"),
+			},
+			want: mustNewCIDRs("10.10.1.0/24", "10.10.3.0/24"),
+		},
+		{
+			name: "keep duplicates",
+			args: args{
+				ipNets:   mustNewCIDRs("10.10.0.0/24", "10.10.1.0/24", "10.10.2.0/24", "10.10.3.0/24", "10.10.0.0/24"),
+				toRemove: mustNewCIDRs("10.10.1.0/24"),
+			},
+			want: mustNewCIDRs("10.10.0.0/24", "10.10.2.0/24", "10.10.3.0/24", "10.10.0.0/24"),
+		},
+		{
+			name: "remove nil",
+			args: args{
+				ipNets:   mustNewCIDRs("10.10.0.0/24", "10.10.1.0/24", "10.10.2.0/24"),
+				toRemove: nil,
+			},
+			want: mustNewCIDRs("10.10.0.0/24", "10.10.1.0/24", "10.10.2.0/24"),
+		},
+		{
+			name: "remove from empty",
+			args: args{
+				ipNets:   []*net.IPNet{},
+				toRemove: mustNewCIDRs("10.10.1.0/24"),
+			},
+			want: []*net.IPNet{},
+		},
+		{
+			name: "remove from nil",
+			args: args{
+				ipNets:   nil,
+				toRemove: mustNewCIDRs("10.10.1.0/24"),
+			},
+			want: nil,
+		},
+	}
+	for _, tt := range tests {
+		result := RemoveAll(tt.args.ipNets, tt.args.toRemove)
+		c.Assert(result, checker.DeepEquals, tt.want, check.Commentf("Test Name: %s", tt.name))
```

Comment:
nitpicking: as I understand it `RemoveAll` modify its first slice parameter by design, and if correct then I think this behaviour should be tested as well.

>  		c.Assert(tt.args.ipNets, checker.DeepEquals, tt.want, check.Commentf("Test Name: %s", tt.name))

Reply:
Yes, `RemoveAll` clobbers its first argument (good point, I should mention that in the doc comment) and must not be used after calling the function - but I don't want to guarantee that the input argument equals the result (similar to e.g. how `append` works) - a future implementation might optimize the by re-allocating and shrinking the capacity of the resulting slice. In that case, the return value will be different from the input argument.